File: test_colors.mjs
import fs from 'fs';
import ts from 'typescript';

const srcCode = fs.readFileSync('./app/water-test.tsx', 'utf8');
const engineSrc = fs.readFileSync('./src/utils/WaterEngine.ts', 'utf8');
const trSrc = fs.readFileSync('./src/utils/terrain.ts', 'utf8');

const merged = `
${trSrc.replace(/export /g, '')}
${engineSrc.replace(/export /g, '').replace(/import .*;/g, '')}
const TEST_WORLD_Z = -100;
const WATER_SIZE = 160;
const size = 160;
const half = 80;

function initRealTerrain(engine) {
    for (let i = 0; i < size * size; i++) {
        const x = i % size;
        const z = Math.floor(i / size);
        const wx = engine.originX - half + x;
        const wz = engine.originZ - half + z;
        engine.T_base[i] = getBaseTerrainHeight(wx, wz);
    }
}

const engineA = new WaterEngine();
engineA.size = size; engineA.originX = 0; engineA.originZ = TEST_WORLD_Z;
engineA.W = new Float32Array(size * size);
engineA.T = new Float32Array(size * size);
engineA.T_base = new Float32Array(size * size);
initRealTerrain(engineA);

const offsetsA = {};
for(let x=-20; x<=20; x++) { offsetsA[\`\${x},\${TEST_WORLD_Z}\`] = 6; offsetsA[\`\${x},\${TEST_WORLD_Z+1}\`] = 6; } 
engineA.updateTerrain([], [], offsetsA);

let coloredPixels = 0;
for(let i=0; i<size*size; i++) {
    const h = engineA.T[i];
    const hBase = engineA.T_base[i];
    const diff = h - hBase;
    
    if (diff > 0.5) coloredPixels++;
}
console.log("Colored pixels in Dam A:", coloredPixels);
`;

const result = ts.transpileModule(merged, { compilerOptions: { module: ts.ModuleKind.CommonJS }});
fs.writeFileSync('test_colors_compiled.js', result.outputText);

File: test_colors.py
import math

size = 160
half = 80
TEST_WORLD_Z = -100

originX = 0
originZ = TEST_WORLD_Z

offsetsA = {}
for x in range(-20, 21):
    offsetsA[f"{x},{TEST_WORLD_Z}"] = 6
    offsetsA[f"{x},{TEST_WORLD_Z+1}"] = 6

coloredPixels = 0

hasOffsets = len(offsetsA) > 1

for i in range(size * size):
    x = i % size
    z = i // size
    wx = originX - half + x
    wz = originZ - half + z
    
    h = 0
    if hasOffsets:
        x0 = math.floor(wx); x1 = x0 + 1
        z0 = math.floor(wz); z1 = z0 + 1
        tx = wx - x0; tz = wz - z0
        
        v00 = offsetsA.get(f"{x0},{z0}", 0)
        v10 = offsetsA.get(f"{x1},{z0}", 0)
        v01 = offsetsA.get(f"{x0},{z1}", 0)
        v11 = offsetsA.get(f"{x1},{z1}", 0)
        
        nx0 = v00 * (1 - tx) + v10 * tx
        nx1 = v01 * (1 - tx) + v11 * tx
        
        h += nx0 * (1 - tz) + nx1 * tz
        
    diff = h - 0
    if diff > 0.5:
        coloredPixels += 1

print(f"Colored pixels in Dam A: {coloredPixels}")
